deepinv/physics/functional/hist.py

```diff
@@ -0,0 +1,294 @@
+
+        if upp is None:
+            upp = x.max(dim=0).values
+    elif torch.is_tensor(edges):
+        edges = edges.flatten().to(x)
+        bins = edges.numel() - 1
+        low = edges[0]
+        upp = edges[-1]
+    else:
+        edges = [e.flatten() for e in edges]
+        bins = [e.numel() - 1 for e in edges]
+        low = [e[0] for e in edges]
+        upp = [e[-1] for e in edges]
+
+        pack = x.new_full((D, max(bins) + 1), float("inf"))
+
+        for i, e in enumerate(edges):
+            pack[i, : e.numel()] = e.to(x)  # pad with inf
+
+        edges = pack
+
+    bins = torch.as_tensor(bins, dtype=torch.long, device=x.device).squeeze()
+    low = torch.as_tensor(low, dtype=x.dtype, device=x.device).squeeze()
+    upp = torch.as_tensor(upp, dtype=x.dtype, device=x.device).squeeze()
+
+    assert torch.all(
+        upp > low
+    ), "The upper bound must be strictly larger than the lower bound"
+
+    if weights is not None:
+        weights = weights.flatten()
+
+    # Filter out-of-bound values
+    if not bounded:
+        mask = ~out_of_bounds(x, low, upp)
+
+        x = x[mask]
+
+        if weights is not None:
+            weights = weights[mask]
+
+    # Indexing
+    if edges is None:
+        idx = quantize(x, bins, low, upp)
+    elif edges.dim() > 1:
+        idx = torch.searchsorted(edges, x.t().contiguous(), right=True).t() - 1
+    else:
+        idx = torch.bucketize(x, edges, right=True) - 1
+
+    idx = torch.clip(idx, min=None, max=bins - 1)  # last bin includes upper bound
+
+    # Histogram
+    shape = torch.Size(bins.expand(D).tolist())
+
+    if sparse:
+        if weights is None:
+            idx, values = torch.unique(idx, dim=0, return_counts=True)
+        else:
+            idx, inverse = torch.unique(idx, dim=0, return_inverse=True)
+            values = weights.new_zeros(len(idx))
+            values = values.scatter_add(dim=0, index=inverse, src=weights)
+
+        hist = torch.sparse_coo_tensor(idx.t(), values, shape)
+        hist._coalesced_(True)
+    else:
+        if D > 1:
+            idx = ravel_multi_index(idx, shape)
+        hist = idx.bincount(weights, minlength=shape.numel()).reshape(shape)
+
+    return hist
+
+
+def histogram(
+    x: Tensor,
+    bins: int = 10,
+    low: float = None,
+    upp: float = None,
+    **kwargs,
+) -> Tensor:
+    r"""Computes the histogram of a tensor.
+
+    This is a `torch` implementation of `numpy.histogram`.
+
+
+    :param torch.Tensor x: A tensor, (*,).
+    :param int bins: The number of bins.
+    :param float low: The lower bound. If `low` is `None` the min of `x` is used instead.
+    :param float upp: The upper bound. If `upp` is `None` the max of `x` is used instead.
+    :param kwargs: Keyword arguments passed to `histogramdd`.
+
+    :return torch.Tensor: The histogram
+    """
+
+    return histogramdd(x.unsqueeze(-1), bins, low, upp, **kwargs)
+
+# if __name__ == "__main__":  # bad practice
```

Comment:
Can we remove?

Reply:
It was just for some benchmarks but yes we can remove it.